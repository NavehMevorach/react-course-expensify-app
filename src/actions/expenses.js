import uuid from 'uuid'


// ------- Expenses Actions ---------

// Add expense
export const addExpense = ({
  title = '',
  note = '',
  amount = 0,
  createdAt = 0
} = {}) => ({
  type: 'ADD_EXPENSE',
  expenses: {
    id: uuid(),
    title,
    note,
    amount,
    createdAt
  }
})

// Remove expense
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
})
// Edit expense

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})
