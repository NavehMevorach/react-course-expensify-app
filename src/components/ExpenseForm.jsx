import React, { Component } from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css';


export class ExpenseForm extends Component {
    constructor(props) {
        super(props)
    
        this.state = {
            title: props.expense ? props.expense.title : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? props.expense.amount : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            focused: false,
            error: ''
        }
    }

    handleChange = (e) => {
        const { name, value } = e.target
        if (name === 'amount') {
            if (value.match(/^\d*(\.\d{0,2})?$/)) {
                this.setState(() => (
                    {
                        [name]: value
                    }
                ))
            }
        } else {
            this.setState(() => (
                {
                    [name]: value
                }
            ))
        }
    }

    onDateChange = (createdAt) => {
        createdAt && this.setState(() => (
            {
                createdAt
            }
        ))

    }

    onFocusChange = ({ focused }) => {
        this.setState(() => (
            {
                focused
            }
        ))
    }

    onSubmit = (e) => {
        e.preventDefault()
        if (!this.state.title || !this.state.amount) {
            this.setState(() => ({ error: 'The form in invalid please fill all the requirement field' }))
        } else {
            this.setState(() => ({ error: '' }))
            this.props.onSubmit({
                title: this.state.title,
                note: this.state.note,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
            })
        }
    }

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>} 
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Title"
                        autoFocus
                        value={this.state.title}
                        onChange={this.handleChange}
                        name='title' />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.handleChange}
                        name='amount' />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.focused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false} />
                    <textarea
                        name="note"
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.handleChange}>
                    </textarea>
                    <button>Add Exchange</button>
                </form>
            </div>
        )
    }
}

export default ExpenseForm
