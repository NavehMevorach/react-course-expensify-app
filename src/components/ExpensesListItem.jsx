import React from 'react'
import { Link } from "react-router-dom"
import moment  from 'moment'
import numeral from 'numeral'

const ExpensesListItem = ({title, note, createdAt, amount, id, dispatch}) => (
    <>
        <Link exact="true" to={`/edit/${id}`}><h3>Title: {title}</h3></Link>
        <p>
        Note: {note}
        Amount: {numeral(amount / 100).format('$0,0.00')}
        Created at: {moment(createdAt).format('MMM Do, YYYY')}</p>

    </>
)

export default ExpensesListItem
