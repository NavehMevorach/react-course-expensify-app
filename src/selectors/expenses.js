import moment from 'moment'

// Filter and Sort  Visible Expenses 
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter(expense => {
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true 
        const endStartMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
        const textMatch = expense.title.toLowerCase().includes(text.toLowerCase())

        return startDateMatch && endStartMatch && textMatch
    }).sort((a,b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        } else {

        }
    })
}