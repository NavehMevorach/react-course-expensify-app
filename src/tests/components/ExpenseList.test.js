import React from 'react'
import { shallow } from 'enzyme'
import { ExpensesList } from '../../components/ExpensesList'
import expenses from '../fixtures/expenses'

test('should render ExpenseList with expense', () => {
    const wrapper = shallow(<ExpensesList expenses={expenses}/>)
    expect(wrapper).toMatchSnapshot()
})


