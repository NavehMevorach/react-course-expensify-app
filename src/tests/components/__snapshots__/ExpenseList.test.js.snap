// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with expense 1`] = `ShallowWrapper {}`;
